### ISSUE TITLE
Is majority voting used during training, or just evaluation?

### ISSUE THREAD

**Issue by ketan0** [adbailey1/DepAudioNet_reproduction#3]
Hi— thanks for making this implementation available! 

I was wondering whether majority voting is used in some way during training, or is it purely for evaluation purposes? I couldn't seem to figure this detail out from the Depaudionet paper. I see that in your `get_output_from_model` function, you say "the output processing depends also on whether soft majority vote or hard majority vote are selected," but it's not apparent to me where majority voting is happening, nor where the option is to switch between soft/hard vote. I see that on line 1203 of `main1.py`, majority voting is calculated on the test set only. Would appreciate if you could shed some light on this. 🙂

**Comment by adbailey1** [comment]
Hi ketan0, 

No problem!

Thanks for this question, I will need to add a clarification for the next update! 

So from the paper, I tried multiple things during the process of building this framework but I settled on only applying majority vote during the inference step (testing on the validation set or the test set). You can see how I perform majority vote in the `def evaluate()` function. Line 204, I get the model's predicted outputs from the inference set. Lines 209-224, I round each prediction and add it to the other predictions from that same folder and for each folder make a count of how many instances of that folder exist. Finally Lines 226-234, I divide the accumulated output of every prediction for each folder by the number of instances for that folder. 

For example: Folder 0 has 5 instances, I round every instance and add together, [0.2, 0.2, 0.2, 0.6, 0.6] goes to [0, 0, 0, 1, 1], therefore accumulated output= 2 and counter = 5. 2/5 = 0.4. This output then is passed to the `def prediction_and_accuracy()` function as per usual where the output is rounded again to give 0 as the final prediction. Hopefully that makes sense, essentially we are just taking the mean of the rounded predictions and then we round that value once more for our final prediction. 

Now, I used to have an option in the config file regarding the type of majority vote (soft or hard) and this basically chose where we applied the majority vote, prior to rounding (soft) or post (hard). I selected hard as the default and removed soft because for extreme cases, the output will not be the majority vote. For example if your predictions are (0.001, 0.55, 0.55), taking the mean of this would result in 0.37, which will give 0 when rounded. Feel free to add this sort of thing into your own code for experimentation though! 

So, I will need to update `def get_output_from_model () `as it no longer has the option for hard or soft MV and I will also look at changing line 1203 because again, there is no implementation of soft or hard MV in my framework. 

Hope this makes sense, and thanks for the question. If you have any more questions feel free to ask. 

Andrew

**Comment by ketan0** [comment]
Great, thank you for the comprehensive answer! That makes sense to me.
